src/main/java/org/apache/commons/collections4/iterators/CartesianProductIterator.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.collections4.iterators;
+
+import java.util.*;
+
+/**
+ * This iterator creates a Cartesian product of the input enumerables,
+ * equivalent to nested for-loops.
+ * <p>
+ * The {@code remove()} operation is not supported, and will throw an
+ * {@code UnsupportedOperationException}.
+ *
```

Comment:
This class could do with some javadoc to state the expected output behaviour, e.g. the iterator creates the Cartesian product of the iterables as a list. Iterables provided to the constructor are used in reverse order, each until exhaustion before proceeding to the next element of the prior iterable and repeating:
```
iterable1 = A, B, C
iterable2 = 1, 2, 3
output:
A, 1
A, 2
A, 3
B, 1
B, 2
...
```